numpy/lib/tests/test_nanfunctions.py

```diff
@@ -388,6 +410,101 @@ def test_empty(self):
             assert_equal(res, tgt)
 
 
+class TestNanFunctions_CumSumProd(TestCase, SharedNanFunctionsTestsMixin):
+
+    nanfuncs = [np.nancumsum, np.nancumprod]
+    stdfuncs = [np.cumsum, np.cumprod]
+
+    def test_allnans(self):
+        # Check for FutureWarning
+        for f, tgt_value in zip(self.nanfuncs, [0, 1]):
+            with warnings.catch_warnings(record=True) as w:
```

Comment:
@shoyer see gh-7099, the assertion context managers are probably nice, especially as context manager. But they use `catch_warnings` inside, and `catch_warnings` is not as helpful as it could be. In gh-7099 I propose a pimped up `catch_warnings` context which actually makes things easier (nested catch warnings, silence warning without catching, still printing warning in release mode). The assertion context managers could use that one internally to be more specific.

Overall, I want our tests to be able to reliably test all warnings in all python versions, and creating a new fancy context manager, while annoying, seemed like the easiest approach.